Spring MVC 활용 Todo API 구현

MJbae opened the pull request:
안녕하세요~ 맥스웰입니다!

지난주에 자바로 Todo API를 구현하면서 정말 많은 것을 배웠고 성장했다고 생각합니다.

이번주에도 열심히 해보겠습니다~

감사합니다!

MJbae committed:
feat: Task 객체에 대한 인터페이스 정의

MJbae committed:
feat: TaskRepository 객체에 대한 인터페이스 정의

MJbae committed:
feat: RepositoryOutput 객체에 대한 인터페이스 정의

MJbae committed:
feat: TaskController 객체에 대한 인터페이스 정의

MJbae committed:
feat: ControllerOutput 객체에 대한 인터페이스 정의

MJbae committed:
chore: 인터페이스로 정의된 객체에 대한 빈 클래스 파일 생성

MJbae commented:
작업내용 
- 절차지향 패러다임의 자료구조 레거시를 의식적으로 제거하고, 객체지향적으로 객체의 인터페이스를 정의했습니다.
- 객체의 메소드는 새로운 것을 만들어서 반환하는 빌더(builder)와 객체의 상태를 수정하는 조정자(manipulator)로 구분하여 정의했습니다.
- 빌더와 조정자의 역할을 분리하기 위해 조정자 메소드는 Output이라는 이름의 객체에 정의했습니다. Respository와 Controller의 조정자 메소드는 각각 RepositoryOutput과 ControllerOuput이라는 별개의 객체에 정의했습니다.

지난주 과제를 수행하고 좋은 리뷰를 받으면서 객체지향 패러다임을 제대로 공부하고 싶다는 욕망이 타올랐습니다.
주말에 '엘레강트 오브젝트', '객체지향의 사실과 오해'를 읽으면서 객체지향적으로 클래스를 정의하는 방법에 대해 공부해봤습니다.
이번 과제에는 주말에 공부한 내용을 토대로, 과제의 요구사항을 객체지향적인 방식으로 설계한 인터페이스에 녹여내는데 집중했습니다.

MJbae committed:
chore: 불필요한 코드 제거

MJbae committed:
refactor: create(조정자) 메소드와 합을 맞추기 위해 responseDtoCreated로 이름 변경

MJbae committed:
refactor: 인터페이스에 정의된 Task.id의 타입을 long으로 통일

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/interfaces/ITaskRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.List;
+
+public interface ITaskRepository {
```
`I`는 제거하는 편이 좋을 것 같습니다!
자바의 기본 API들을 보면 인터페이스들에 `I` 를 붙이지는 않고 있는 모습이 보입니다. (`List`, `Map`, `Iterable`)

MJbae replied:
그렇군요! 자바 공식문서를 보니 Interface 작명 컨벤션에는 I를 붙인다는 내용이 없네요!
알고보니 Interface 이름에 I를 붙이는 건 C#의 컨벤션이네요 ㅎㅎ

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/RepositoryOutput.java:
```diff
@@ -0,0 +1,4 @@
+package com.codesoom.assignment.controllers;
+
+public class RepositoryOutput {
```
이건 어떤 용도의 클래스인지 궁금하군요! 다음 PR이 기대됩니다!

MJbae replied:
- RepositoryOutput의 역할은 Repository 객체에서 관리하는 데이터에 대한 갱신(Create, Update, Delete)입니다! 

- 주말에 '엘레강트 오브젝트'라는 책에서 얻은 아이디어를 토대로 구현해봤습니다. 책에 따르면 객체의 책임을 직관적으로 인지하기 위해 메소드를 빌더와 조정자로 구분하여 정의할 것을 권장하고 있습니다. 이에 따라 Repository에는 빌더 메소드만 정의하고 RepositoryOutput에는 조정자 메소드를 정의했습니다. 

- 책에서 정의하는 빌더(builder) 메소드의 책임은 새로운 것을 만들어서 반환하는 것이고, 조정자(manipulator) 메소드의 책임은 객체의 상태를 변경하는 것입니다.

MJbae committed:
fix: Task 객체의 id를 HashMap의 키로 사용하기 위해 wrapper class로 수정

MJbae committed:
refactor: Java의 Interface 이름 컨벤션을 따르기 위해 인터페이스의 첫글자 'I' 제거

MJbae committed:
refactor: 적합한 경로에서 관리하기 위해 RepositoryOutput 객체의 위치를 TaskRepository와 같은 경로로 이동

MJbae committed:
refactor: 자바 인터페이스 작명 컨벤션에 따라 인터페이스의 이름을 보다 추상적인 이름으로 설정

MJbae committed:
feat: DefaultTask 인터페이스에 따라 Task 클래스 정의

MJbae committed:
feat: Task 객체 간 id를 기준으로 내림차순 정렬하기 위해 Comparable 인터페이스 구현

MJbae committed:
feat: DefaultRepository 인터페이스에 따라 TaskRepository 구현

MJbae committed:
feat: DefaultController 인터페이스에 따라 TaskController 구현

MJbae committed:
feat: 유지보수성을 높이기 위해 Task 객체 처리에 따라 HTTP Response에 필요한 부분만 TaskResponseDto로 정의

MJbae committed:
feat: TaskController와 TaskRepository를 Spring Bean으로 등록하기 위해 component annotation 추가

MJbae committed:
feat: 스프링 어플리케이션 동작을 위해 SpringApplication::run 추가

MJbae commented:
안녕하세요 영환님
지난 리뷰 감사합니다. 덕분에 자바의 인터페이스 컨벤션에 대해 더 공부할 수 있었습니다.

이번 작업의 주요내용은 '인터페이스 보강'과 'GET 요청 처리로직 구현'입니다.
- 인터페이스의 경우 자바 문서의 작명 컨벤션에 따라 구현부의 클래스보다 추상화하기 위해 Default를 이름 앞에 붙였습니다. Default를 붙인 이유는 해당 인터페이스들이 MVC 패턴을 따르되 Todo List의 가장 기본이 되는 형태를 띄기 때문입니다.
- 앞서 정의한 인터페이스에 따라 TaskController ,Task, TaskRepository 및 기타 DTO를 정의했습니다.

MJbae committed:
refactor: RepositoryOutput의 구현체 이름을 구체적으로 수정(RepositoryOutput -> TaskRepositoryOutput)

MJbae committed:
feat: RepositoryOutput 인터페이스를 따라서 TaskRepositoryOutput 클래스 구현

MJbae committed:
feat: DTO에서 Task로 변환하기 위해 public 생성자 추가

MJbae committed:
feat: ControllerOutput 구현

MJbae committed:
feat: TaskControllerOutput 구현

MJbae committed:
feat: TaskRequestDtoCreating 구현

MJbae committed:
feat: TaskRequestDtoUpdating 구현

MJbae committed:
fix: 직렬화를 위해 기본 생성자 및 getter 추가

MJbae committed:
feat: Http Method 및 상태코드 맵핑 어노테이션 추가

MJbae commented:
안녕하세요 영환님
지난 리뷰 감사합니다. 덕분에 자바의 인터페이스 컨벤션에 대해 더 공부할 수 있었습니다.

- 인터페이스의 경우 자바 문서의 작명 컨벤션에 따라 구현부의 클래스보다 추상화하기 위해 Default를 이름 앞에 붙였습니다. Default를 붙여서 DefaultRepository 등을 정의한 이유는 해당 인터페이스들이 MVC 패턴을 따르되 Todo List의 가장 기본이 되는 인터페이스가 되기 때문입니다.
- 앞서 정의한 인터페이스에 따라 TaskController ,Task, TaskRepository 및 기타 DTO를 정의했습니다.

MJbae committed:
refactor: Ouput 클래스에 DI 적용하기 위해 인스턴스 객체를 반환하도록 수정

MJbae commented:
안녕하세요 영환님
지난 리뷰 감사합니다. 덕분에 자바의 인터페이스 컨벤션에 대해 더 공부할 수 있었습니다.

주요작업
- 인터페이스의 경우 자바 문서의 작명 컨벤션에 따라 구현부의 클래스보다 추상화하기 위해 Default를 이름 앞에 붙였습니다. Default를 붙여서 DefaultRepository 등을 정의한 이유는 해당 인터페이스들이 MVC 패턴을 따르되 Todo List의 가장 기본이 되는 인터페이스가 되기 때문입니다.
- 앞서 정의한 인터페이스에 따라 TaskController ,Task, TaskRepository를 구현했습니다.
- DTO의 경우 직렬화 역직렬화의 기능을 구현하던 중 자바빈의 레거시인를 만족시키기 위해 기본 생성자와 getter를 추가했습니다.

MJbae committed:
fix: 요구사항에 따라 Task에 대한 수정 요청은 id를 pathparam으로 전달하도록 인터페이스 수정

MJbae committed:
fix: 인터페이스 수정에 따라 TaskControllerOupt::update 시그니처 수정

MJbae committed:
refactor: TaskRequestDtoCreating과 TaskRequestDtoUpdating을 TaskRequestDto로 통합

MJbae committed:
feat: PathVariable 어노테이션 추가

MJbae committed:
refactor: 매직넘버 1L을 상수로 변경

MJbae committed:
refactor: 객체 간 결합도를 낮추기 위해 클래스 멤버 변수 수정

작업내용
- TaskRepository와 다른 객체 간 결합도를 낮추기 위해서 생성자 내부의 인스턴스 생성 로직을 클래스 멤버 변수 선언부로 이동
- 전역변수는 객체 간 데이터를 주고 받는 객체지향 패러다임에 어울리지 않으므로 제거

MJbae committed:
refactor: Task.id의 시작점을 의미하는 변수의 의미에 부합하는 이름으로 수정

MJbae commented:
**객체지향 패러다임과 스프링의 동작원리를 고려하여 모든 클래스 멤버 변수에서 static 키워드를 제거했습니다.**
- 전역변수는 객체 간의 협력을 무시하고 접근할 수 있는 여지를 남기기 때문에 객체 간 결합도를 높이는 부작용으로 이어질 수 있습니다
- 스프링에서는 Component 어노테이션이 있는 객체가 일종의 '전역 인스턴스 객체'(싱글톤)로 관리되기 때문에 클래스 멤버 변수를 별도의 전역변수로 관리할 필요가 없습니다.

**Request DTO에서 불변 객체인 Entity로 변환하는 부분에서 의견을 듣고 싶습니다.**
- DTO 클래스 RequestDto에서 toEntity 메소드를 활용해서 Entity 객체인 Task를 생성하고자 합니다. Task 클래스가 불변 객체로 정의된다면, DTO 클래스로부터 전달 받지 못한 id를 Task 클래스에 어떻게 주입할 수 있을까요?
- POST 요청에 따라 Request Body에는 할 일의 제목(title)만 전달되기 때문에 DTO의 속성값에는 title만 존재합니다.
- Repository에서 Task의 Id 값이 자동증가시키는 역할을 하고 있는데, Repository로부터 자동증가된 Task Id를 DTO로 가져오자니 계층을 너무 뛰어넘는 것 같아서 관련 작업을 보류한 상태입니다.

MJbae committed:
clean: 불필요한 코드 제거

MJbae committed:
refactor: Task 클래스를 불변객체로 만들기 위한 작업

작업내용
- Task 내 필드를 수정하는  메소드 제거
- Task 필드를 수정하는 로직을 Task 생성자로 대체

MJbae committed:
feat: TaskResponseDto에 Id 속성값 처리 부분 구현

MJbae committed:
feat: POST, PUT 요청에 따라 객체를 TaskResponseDto로 반환하는 로직 생성

MJbae committed:
feat: 스프링 제공 RequestMapping 기능을 활용하기 위한 타협

작업내용
- responseDto를 반환하는 로직을 Request를 받아서 처리하는 메소드에 통합

MJbae committed:
feat: Path Param으로 전달 받은 Task Id 유효성 검사 로직 추가

작업내용
- Repository 객체 내부에 매개변수로 전달된 Id에 해당하는 Task 객체가 존재하는 지 확인하는 메서드 추가
- Task Id에 맞는 Task 인스턴스가 없을 시 NotFound Response를 전달하는 로직 추가

MJbae committed:
chore: package-lock.json 업로드

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskRepositoryOutput.java:
```diff
@@ -0,0 +1,47 @@
+    private Long savedTaskId;
+    private Long updatedTaskId;
+
+    TaskRepositoryOutput(Map<Long, Task> tasks, Long sequence) {
+        this.tasks = tasks;
+        this.sequence = sequence;
+    }
```
접근 제어자가 없는 것은 의도된 것일까요?

MJbae replied:
리뷰 감사합니다! public 클래스의 인스턴스 생성은 모두 생성자를 거치므로 생성자의 접근제어자는 package-private이 아니라 public이 적절하겠네요!

ssisksl77 commented:
> **객체지향 패러다임과 스프링의 동작원리를 고려하여 모든 클래스 멤버 변수에서 static 키워드를 제거했습니다.**
> 
> * 전역변수는 객체 간의 협력을 무시하고 접근할 수 있는 여지를 남기기 때문에 객체 간 결합도를 높이는 부작용으로 이어질 수 있습니다
> * 스프링에서는 Component 어노테이션이 있는 객체가 일종의 '전역 인스턴스 객체'(싱글톤)로 관리되기 때문에 클래스 멤버 변수를 별도의 전역변수로 관리할 필요가 없습니다.
> 
> **Request DTO에서 불변 객체인 Entity로 변환하는 부분에서 의견을 듣고 싶습니다.**
> 
> * DTO 클래스 RequestDto에서 toEntity 메소드를 활용해서 Entity 객체인 Task를 생성하고자 합니다. Task 클래스가 불변 객체로 정의된다면, DTO 클래스로부터 전달 받지 못한 id를 Task 클래스에 어떻게 주입할 수 있을까요?
> * POST 요청에 따라 Request Body에는 할 일의 제목(title)만 전달되기 때문에 DTO의 속성값에는 title만 존재합니다.
> * Repository에서 Task의 Id 값이 자동증가시키는 역할을 하고 있는데, Repository로부터 자동증가된 Task Id를 DTO로 가져오자니 계층을 너무 뛰어넘는 것 같아서 관련 작업을 보류한 상태입니다.

- 기존 `Task` 에 `id` 를 주입한다에서 `id` 가 있는 `Task` 객체를 만든다고 생각해보면 어떨까요!
- 좋습니다! `Id를 만들어주는 역할` 을 하는 객체를 만들어서 분리해보면 무엇할 수 있을지 더 명확하게 알 수 있을지 않을까요?!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/Task.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public int compareTo(Task task) {
+        if (id() == null || task.id() == null) {
+            return 0;
+        }
```
id 둘중 하나가 `null` 이면 같은 것으로 간주하는 것은 아주 중요한 정책인 것 같아요!

해당 방식이 과연 옳은지 고민해보고 옳다면 주석을 추가해주세요!

MJbae replied:
IDE에서 자동완성해준 코드를 살펴보지 않으니 이런 참사가 발생하는군요! 주옥 같은 리뷰 감사합니다.
Task 객체는 불변객체이기 때문에 id가 null일 수 없습니다. 혹여나 비교 대상이 되는 id 중 하나가 null이라도 두 객체가 순서가 같다는 건 전혀 의도되지 않은 부분입니다. 불필요한 조건식이기 때문에 제거하는 게 맞겠네요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/interfaces/RepositoryOutput.java:
```diff
@@ -0,0 +1,17 @@
+    void deleteBy(Long id);
+
+
+    Task taskSaved();
```
taskSaved는 가장 최근에 저장한 값을 가져오는 군요! 하지만 방금 제가 저장한 `Task`임을 보장하진 못하는 것 같아요!

MJbae replied:
@ssisksl77 맞습니다! 현재 로직으로는 외부 요인에 따라 동작하지 않을 수 있겠네요!
Controller의 HTTP 요청 처리 로직에 대해 Transaction을 거는 방법으로 문제를 해결할 수 있겠네요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/interfaces/DefaultTask.java:
```diff
@@ -0,0 +1,7 @@
+package com.codesoom.assignment.interfaces;
+
+public interface DefaultTask {
```
인터페이스를 사용하는 다른 개발자분들은 인터페이스를 보고 어떤 역할을 하는지 알 거에요! 한번 주석을 남겨주세요!

MJbae replied:
오늘은 시간이 부족해서 주석을 남기는 작업은 금요일에 해보겠습니다!

MJbae replied:
> 인터페이스를 사용하는 다른 개발자분들은 인터페이스를 보고 어떤 역할을 하는지 알 거에요! 한번 주석을 남겨주세요!

인터페이스에 주석을 남겼습니다. 구현체에도 주석을 남겨야하나 고민했지만, 인터페이스만 보더라도 쉽게 이해할 수 있는 간단한 API이기 때문에 구현체에는 주석을 생략했습니다.

MJbae replied:
주석을 작성하는 과정에서 객체에 대한 이름이 막 떠오르네요~ 앞으로 인터페이스 만큼은 설계 단계에서 꼭 주석을 남겨보겠습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/interfaces/RepositoryOutput.java:
```diff
@@ -0,0 +1,17 @@
+
+import com.codesoom.assignment.Task;
+
+public interface RepositoryOutput {
```
`RepositoryOutput` 이라는 이름에서 `Output` 이 들어가면 무엇을 더 기대할 수 있는 걸까요? 네이밍에서 `Output` 의 의미를 자세히 남겨주세요!

MJbae replied:
역할에 따라서 이름을 아래와 같이 수정해보았습니다.
- TaskController -> TaskShowingController
- TaskControllerOutput -> TaskManipulatingController
- TaskRepository.java -> TaskLoadingRepository
- TaskRepositoryOutput -> TaskManipulatingRepository

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/dtos/TaskRequestDto.java:
```diff
@@ -0,0 +1,22 @@
+public class TaskRequestDto {
+    private String title;
+
+    TaskRequestDto() {
```
`title` 이 없는 `TaskRequestDto` 도 유효한 걸까요? 한번 인자가 없는 생성자로 `@RequestBody` 에서 동작하도록 개선해보아요!

MJbae replied:
@ssisksl77 title이 없으면 유효한 DTO로 볼 수 없다는 방향으로 개선했습니다.
스프링에서 제공하는 어노테이션이 아니라 유효성검사의 역할을 수행하는 객체를 정의했습니다.
이어서 Controller 내 요청을 처리하는 메소드에서 유효성 검사 객체의 인스턴스를 생성하는 방향으로 작업했습니다.

MJbae committed:
refactor: public 클래스의 생성자에 public 접근제어자 추가

MJbae committed:
refactor: Comparable::compareTo 내 불필요한 비교식 제거

MJbae committed:
fix: requestDto에 대한 유효성 검증 로직 추가

작업내용
- TaskRequestDto에 title 필드가 null or "" or " " 일 경우, 400 상태로 응답함

MJbae committed:
refactor: RequestDto와 Path Param에 대한 유효성 검사 로직을 private mehthod로 분리

MJbae committed:
refactor: request param과 request body에 대한 유효성 검사를 수행하는 객체 생성 및 적용

작업내용
- Controller 내 메소드로 유효성 검사하던 로직을 독립적인 객체로 정의
- 유효성검사 객체를 TaskController와 TaskControllerOuput에 적용

MJbae committed:
fix: 외부 요인에 의한 갱신 작업 이슈를 해결하기 위해 Task 생성 및 수정 요청에 메서드에 Transactional 추가

MJbae committed:
chore: Data Source 미설정 상태에서 정상 빌드되도록 관련 설정 적용

MJbae committed:
feat: CORS 정책에 따라 ORIGIN response header fields를 설정하기 위해 CrossOrigin 어노테이션 추가

MJbae committed:
refactor: Task 객체를 불변으로 만들기 위해 Id 주입 방식 변경

작업내용
- title만으로 Task 객체를 생성하더라도 UUID 활용하여 고유의 아이디를 넣어서 Task 객체를 추가하도록 수정

MJbae committed:
fix: Path Parm으로 전달될 수 있는 값의 범위를 고려하여 UUID max 값을 integer max값을 기준으로 설정

MJbae committed:
refactor: Output이라는 이름을 역할과 연계하여 수정

작업내용
- TaskController -> TaskShowingController
- TaskControllerOutput -> TaskManipulatingController
- TaskRepository.java -> TaskLoadingRepository
- TaskRepositoryOutput -> TaskManipulatingRepository

MJbae commented:
> > **객체지향 패러다임과 스프링의 동작원리를 고려하여 모든 클래스 멤버 변수에서 static 키워드를 제거했습니다.**
> > 
> > * 전역변수는 객체 간의 협력을 무시하고 접근할 수 있는 여지를 남기기 때문에 객체 간 결합도를 높이는 부작용으로 이어질 수 있습니다
> > * 스프링에서는 Component 어노테이션이 있는 객체가 일종의 '전역 인스턴스 객체'(싱글톤)로 관리되기 때문에 클래스 멤버 변수를 별도의 전역변수로 관리할 필요가 없습니다.
> > 
> > **Request DTO에서 불변 객체인 Entity로 변환하는 부분에서 의견을 듣고 싶습니다.**
> > 
> > * DTO 클래스 RequestDto에서 toEntity 메소드를 활용해서 Entity 객체인 Task를 생성하고자 합니다. Task 클래스가 불변 객체로 정의된다면, DTO 클래스로부터 전달 받지 못한 id를 Task 클래스에 어떻게 주입할 수 있을까요?
> > * POST 요청에 따라 Request Body에는 할 일의 제목(title)만 전달되기 때문에 DTO의 속성값에는 title만 존재합니다.
> > * Repository에서 Task의 Id 값이 자동증가시키는 역할을 하고 있는데, Repository로부터 자동증가된 Task Id를 DTO로 가져오자니 계층을 너무 뛰어넘는 것 같아서 관련 작업을 보류한 상태입니다.
> 
> * 기존 `Task` 에 `id` 를 주입한다에서 `id` 가 있는 `Task` 객체를 만든다고 생각해보면 어떨까요!
> * 좋습니다! `Id를 만들어주는 역할` 을 하는 객체를 만들어서 분리해보면 무엇할 수 있을지 더 명확하게 알 수 있을지 않을까요?!


유레카! 의견 감사합니다~
- 'id'가 있는 Task를 불변으로 만들어보았습니다. Task 객체 자체적으로 생성자에 title만 입력 받더라도 UUID를 활용하여 유니크한 id를 생성할 수 있도록 수정했습니다.
- 다음에는 Id를 만들어주는 객체로 'Id' 객체를 정의해서 관련 기능을 분리해보겠습니다!

MJbae committed:
feat: PATCH 요청을 처리할 수 있도록 RequestMapping 어노테이션 수정

MJbae committed:
chore: add another gradle dependancy, mavenCentral

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/dtos/TaskResponseDto.java:
```diff
@@ -0,0 +1,24 @@
+import com.codesoom.assignment.Task;
+
+public class TaskResponseDto {
+    private Long id;
+    private String title;
+
+    public TaskResponseDto() {
+    }
```
```suggestion
    private final Long id;
    private final String title;
```

이런 형태가 되도록 개선해보면 좋은 경험이 될 것 같아요!

MJbae replied:
@ssisksl77 ResponseDto는 역직렬화에 사용되므로 빈 생성자가 필요없겠군요! 콕 집어주셔서 감사합니다 :)
RequestDto의 경우에도 불변객체로 만들 수 있는 방법에 대해 찾아보니 딱히 없어보이네요. 대안으로 빈 생성자에 private 접근제어자를 달아서 외부에서 빈 생성자로 객체생성을 막는 방법이 있네요!
출처: https://github.com/woowacourse-moltudy/15-mins-retrospective/issues/25

ssisksl77 replied:
```java
public class TaskRequestDto {
    private String title;

    @ConstructorProperties({"title"})
    private TaskRequestDto(final String title) {
        this.title = title;
    }

    public String getTitle() {
        return title;
    }

    public Task toEntity() {
        return new Task(title);
    }
}
```

이 코드를 한번 참고해주시면 좋겠어요!

MJbae replied:
멋진 노하우 공유해주셔서 감사합니다!!
ConstructorProperties 어노테이션이 있으면 빈 생성자가 아니라 해당 어노테이션이 붙은 생성자를 역직렬화에 사용하네요
와우! 환상적이네요~ 그렇다면 TaskRequestDto도 불변으로 만들 수 있겠군요!

LeeJongbokz replied:
맥스웰님 작성해주신 코드 잘 보았습니다 ㅎㅎ
혹시 
1) TaskRequestDTO와 TaskResponseDTO를 구분하신 이유에 대해 알 수 있을까요?

감사합니다~

MJbae replied:
@LeeJongbokz 요구사항을 분석해보면 Request Body에 들어오는 값은 오직 title 뿐이죠. 하지만 Response Body로 내려줄 때는 title과 id를 함께 보내주어야 합니다. 더불어 Request Body를 직렬화하는 역할과 Response Body로 역직렬화하는 역할에서 일부 기능에 차이가 있습니다. 예를 들어 toEntity라는 메소드는 두 역할 중 한쪽에만 필요한 것이죠~
그래서 구분할 필요가 있다고 판단했습니다 :)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/interfaces/DefaultController.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.util.List;
+
+public interface DefaultController {
```
스프링MVC에서 `Controller` 는 요청에 따라 어떤 객체에게 역할을 넘길 것인지를 담당합니다. 

또한 이 인터페이스는 또 다른 `Controller` 인 `ManipulatingController` 를 갖고 있네요!

왜 `TaskShowingController` 는 `manipulator` 를 갖고 있어야 할까요?

MJbae replied:
CRUD와 연관된 모든 컴포넌트에 대해 R과 CUD를 구분하는 건 불필요하다는 생각이 드네요!
말씀처럼 Controller에서 또다른 컨트롤러를 갖고 있는 것도 불필요하게 분리된 것 같습니다.
Repository는 조회와 갱신을 분리하되 Controller는 합치는 방향으로 작업해보겠습니다.

MJbae replied:
Repository의 경우 엔티티에 대한 조회와 갱신으로 역할을 나눌 수 있지만, Controller의 경우 HTTP Method에 따라 요청을 받아서 그 결과값을 반환해주는 하나의 역할을 수행하는 군요! GET 요청에 대한 작업과 POST 요청에 대한 작업은 결국 '요청 처리 및 처리 값 반환'이라는 같은 역할로 보는게 맞겠네요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/interfaces/DefaultRepository.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.List;
+
+public interface DefaultRepository {
```
`Default` 가 유용할 수도 있지만 여기에는 좀 더 명확한 역할을 적어주면 좋겠어요!

MJbae replied:
@ssisksl77 각 인터페이스의 역할에 대해 더 생각해보니 대체할 수 있는 표현이 있네요!
아래와 같이 수정했습니다. 그런데 DefaultTask 만큼은 Default를 대체할 만한 이름이 떠오르질 않네요
- DefaultRepository -> LoadingRepository
- DefaultController -> CrudController

ssisksl77 replied:
아주 좋아요! Task가 너무 간단하고 아직 Task를 다른 구현체로 만들어볼만한 요구사항이 있지도 않죠!

ssisksl77 replied:
한가지 아쉬운 것은 만약에 `DefaultTask` 와 `Task` 라는 이름을 쓴다하더라도!
`DefaultTask` 가 `interface` 인 것보단 `Task` 라는 이름이 인터페이스 인 것이 더 일반적인 것으로 보여요!

ssisksl77 replied:
또한 `Default` 가 조금 아쉬운 것은 `default` 라는 것은 비즈니스로직에서 자주 바뀔 수 있는 부분이라는 것입니다!

ssisksl77 replied:
`List` 인터페이스를 기억해보세요!
`List`, `ArrayList`, `LinkedList` 에서 인터페이스는 `List` 입니다!

MJbae replied:
맞네요! Task가 더 추상화된 표현이므로 인터페이스의 이름으로 더 적합하겠네요!
크~ 감사합니다~

LeeJongbokz replied:
맥스웰님~
1) DefaultTask 인터페이스를 따로 만드신 이유에 대해서 알 수 있을까요?
2) Repository 인터페이스와 클래스를 만드신 것은 
    실제 지금 DB에 저장하는 것까지 구현을 하신 것일까요?
    
감사합니다ㅎㅎ

MJbae replied:
@LeeJongbokz 좋은 질문 감사합니다. 저는 Interface는 객체 간 협력의 설계도이자 협력의 계약으로 보고 있습니다. Class는 설계도 및 계약에 따른 구현체로 보고 있구요. 그래서 class로 구현하기 전에 객체지향 패러다임에 따라 먼저 객체 간의 협력을 어떻게 이루어 갈 것인지에 대해서 고민한 부분을 인터페이스에 반영했습니다. 모든 class에 대해 interface의 형태로 설계도를 그리는 것 보다 주요 객체인 엔티티(Task), 엔티티 간접 관리(Repository), HTTP Request 처리(Controller)에 대한 역할을 가진 객체에 대해서만 인터페이스를 만들게 되었습니다 :)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/interfaces/DefaultRepository.java:
```diff
@@ -0,0 +1,15 @@
+
+    Task taskBy(Long id);
+
+    boolean notPresent(Long id);
```
`notPresent` 보다는 `isNotPresent` `isPresent` 가 좀 더 직관적이에요!
boolean으로 술어를 만들 때는 `is-` 가 prefix로 붙는 것이 타 개발자에게 좀 더 가독성이 좋을 거에요!

MJbae replied:
그렇군요! 'is'가 있는 메소드 이름에 많은 개발자가 익숙하니 가독성 면에서 다시 생각해볼만 하네요. 찾아보니 java core packages의 대부분의 boolean 반환 메소드의 이름에 is가 붙기도 하는군요.

하지만 저는 elegant object에서 얻은 두 가지 이유로 is를 제거했습니다.
- 'is'가 붙지 않아도 메소드의 책임을 이해할 수 있기 때문입니다. boolean 메소드는 주로 조건문에서 사용된다는 점, 메소드의 이름이 형용사로 끝난다는 점에서 'is'를 제거해도 다른 개발자분들이 메소드의 책임을 충분히 이해할 수 있을 것으로 생각합니다.
- 'is'를 제거함으로써 메소드의 이름이 더 간결해집니다. 사실 'is'를 제거해서 얻는 간결함이 큰 의미가 없을 수 있겠지만, 메소드의 이름으로 메소드의 책임을 충분히 전달한다는 전제하에 이름에서 제거할 수 있는 부분이 있다면 제거하는 것이 낫다고 판단했습니다.
- 더불어 java naming convention 문서(by oracle)을 살펴보니, boolean 반환 메소드에 대해 'is'를 붙이는 것에 대한 컨벤션이 없다는 것도 참고 했습니다.

ssisksl77 replied:
맞아요! `is` 를 붙이는 것이 필수는 아니에요! 법칙은 아니니깐요! 

`boolean` 을 리턴해도 그 의도는 명사만으로는 살~짝 아쉬울 때가 있을거에요! 술어부에서 메소드에서 자주쓰는 접두어는 `is-` , `has-`  가 일반적이죠!  

예를들어 자바의 `Interator` 라는 인터페이스 명세를 보면 `hasNext` 라는 메소드가 있어요! 이것은 `isNext` 와는 전혀 다른 의미죠! 이것만 기억해주시면 됩니다!

또한 충분히 이해가 가능함에는 주석이 필수라고 생각해요! 주석을 꼭 붙여주세요!

MJbae replied:
그렇군요! 이런 포인트에서는 주석을 남길 필요가 있겠네요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskManipulatingRepository.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Override
+    public void update(Long oldTaskId, Task newTask) {
```
여기서 제가 헷갈렸던 점을 말씀드릴게요!
`newTask` 라고 하면 이미 `id`가 있고 이곳에 새로운 `title` 이 있다고 저는 간주하였어요!

그래서 `oldTaskId` 가 왜 필요한가! 에 대해서 고민하게 되었지요. 

구현을 보니 `newTask` 의 `id` 를 쓰지 않고 있는 것을 발견했어요!

이런 저의 생각을 바탕으로 `update` 메소드를 개선해보셨으면 좋겠어요!

MJbae replied:
그렇군요! 말씀해주신 방향으로 작업하면 controller의 역할과 repository의 역할에 어울리는 방향으로 개선되겠네요!
controller에서는 request로 받은 데이터를 토대로 Task를 구성하고, repository에서는 controller에서 받은 Task를 Map에 반영하는 작업만 수행하도록 개선했습니다!

MJbae committed:
refactor: ShowingController와 ManipulatingController를 Controller로 통합

MJbae committed:
refactor: Interface 이름을 Default 대신 역할을 드러낼 수 있는 방향으로 수정

MJbae committed:
refactor: Repository에서 Task 수정 로직에 필요한 매개변수를 최소화하는 방향으로 작업

작업내용
- Controller 계층에서 업데이트할 Task를 생성
- Repository에서 이미 생성된 Task를 받아와서 tasks에 업데이트 하도록 수정

MJbae committed:
refactor: Controller의 구현체와 인터페이스의 역할을 잘 드러내기 위해 이름 변경

MJbae committed:
chore: 인터페이스에 대한 Java doc 추가

MJbae committed:
refactor: TaskRequestDto를 불변객체로 수정

작업내용
- public 기본 생성자 제거
- 필드를 final 키워드로 정의

MJbae committed:
refactor: ObjectMapper에서 사용하는 빈 생성자에 접근제어자를 private으로 설정

MJbae committed:
refactor: Validation 클래스를 validations 경로 이하에서 관리

MJbae committed:
refactor: 예외 메시지에 원인과 해결방법을 명시하는 방향으로 수정

MJbae committed:
refactor: 객체 내부와 객체 간 협력에 사용되는 모든 객체를 불변으로 사용하기 위해 final 키워드 적용

작업내용
- 모든 매개변수 final 키워드 적용
- 모든 지역변수에 final 키워드 적용

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/interfaces/ManipulatingRepository.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Summary:
+ * Task 엔티티 객체에 대한 갱신
+ * <p>
+ * All Known Implementing Classes:
+ * TaskManipulatingRepository
+ * <p>
+ * Method Parameters:
+ * Task task - Task 엔티티 객체
+ * Long id - Task 객체의 Id에 해당하는 객체
+ * <p>
+ * Method Returns:
+ * Task - Task 엔티티 객체
+ */
```
`<p>` 를 쓰셨다면 `</p>` 쌍으로 묶는 것이 좋아요!

인자와 리턴에 대한 설명은 각 메소드에 추가하는 것이 좋겠어요!

javadoc의 맨 위는 어떤 역할을 하는지 설명을 하는 공간이기 때문에 `Summary:` 라는 단어는 없어도 되요!

MJbae replied:
아하! 그렇군요! 자바 주석 컨벤션에 대해 다시 확인해봐야겠네요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/interfaces/TaskController.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.List;
+
+/**
+ * Summary:
+ * Task 객체에 연관된 HTTP CRUD 요청을 받고, 처리결과를 응답으로 반환
+ * <p>
+ * All Known Implementing Classes:
+ * TaskCrudController
+ * <p>
+ * Method Parameters:
+ * TaskRequestDto requestDto - Request Body로 전달된 JSON 객체를 직렬화하여 받기 위한 객체
+ * Long id - Request Path Parameter 전달된 Task Id를 받기 위한 객체
+ * <p>
+ * Method Returns:
+ * TaskResponseDto - HTTP Request를 처리한 결과를 JSON 객체로 역직렬화하기 위한 객체
+ */
```
여기도 위에 설명드린 것과 마찬가지에요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/TaskManipulatingRepository.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Map;
+
+public class TaskManipulatingRepository implements ManipulatingRepository {
+    private final Map<Long, Task> tasks;
+    private Long savedTaskId;
+    private Long updatedTaskId;
```
병렬성을 한번 고려해보면 좋을 것 같아요!

MJbae replied:
@ssisksl77 병렬성 이슈를 해결하기 위해서는 savedTaskId, updatedTaskId 변수를 제거하고 각각 갱신과 반환의 책임을 지는 메소드를 합치는 방법 외에는 떠오르지 않네요.

혹시 TaskManipulatingRepository에서 save 또는 update 메소드가 동작함에 따라서 savedTaskId 또는 updatedTaskId에 Id값이 할당되는 구조에서 병렬성을 함께 가져갈 수 있는 방법이 있을까요?

ssisksl77 replied:
변수를 제거하지 않고 병렬성을 갖고가고 싶으신거군요!
변수가 여러개이면 병렬성을 처리하기 힘들죠. 

일단 저는 `각각 갱신과 반환의 책임을 지는 메소드를 합치는 방법` 이 무엇을 뜻하는지 잘 모르겠습니다. ㅠㅜㅠㅜ 코드로 한 번 보여주셨으면 싶은데 아쉽네요 ㅠㅜㅠㅜ

일단 저는 상태를 줄이는 것이 시작이라고 생각해요! 애초에 병렬성을 가지는데 특정 공통변수를 여러 스레드가 변경을 하는 상황을 만들면 안되겠죠! 

자바에는 concurrent 를 담당하는 api가 있습니다. 거기에 꽤나 유용한 객체들이 많이 있어요! `AtomicLong`, `ConcurrentHashMap` , `ConcurrentLinkedQueue` 등등이 있어요.

가장 간단하게는 `synchronized` 라는 키워드를 이용해서 일련의 작업들을 락으로 막을 수 있어요!

MJbae replied:
그렇군요! 다음에는 커밋에 대한 링크를 참고해서 말씀드리는 편이 좋겠네요. 라스님의 PR에는 코멘트에다가 커밋내역을 링크로 참조할 수 있도록 하던데 따라해보겠습니다! 
참고로 `갱신과 반환의 책임을 지는 메소드를 합치는 방법`은  [이 커밋](https://github.com/CodeSoom/spring-week2-assignment-1/pull/79/commits/97085b3b9b5a6b685ae9abc2abe5faab284b84f3)의 작업 내용이었습니다. 

지바의 병렬성에 대한 부분은 날짜를 잡고 학습해봐야 겠네요. 리뷰로 언급해주신 키워드를 참고하겠습니다. 감사합니다~

MJbae committed:
refactor: TaskExceptionAdvice를 활용하여 task id에 해당하는 task가 없는 경우에 대해 전역적으로 예외 처리

MJbae committed:
chore: 인터페이스에 대한 주석 수정

작업내용
- 불필요한 summary 제거
- 인터페이스 내 정의된 메소드에 대해 주석 추가

MJbae committed:
refactor: Task에 대한 인터페이스와 구현체의 이름 수정

작업내용
- DefaultTask라는 이름을 구현체로 수정
- Task라는 이름을 인터페이스로 수정

사유
- 인터페이스의 이름에 대한 컨벤셩은 구현체 보다 더 추상화되는 것인데 전에는 구현체의 이름이 더 추상성을 가지고 있었기 때문임

MJbae committed:
refactor: 병렬성 이슈를 해결하기 위해 TaskManipulatingRepository에서 갱신과 반환을 분리했던 메소드를 통합

MJbae committed:
refactor: ConstructorProperties 어노테이션 적용하여 RequestDto 객체를 불변으로 수정

작업내용
- 모든 클래스 멤버변수가 초기화되는 생성자에 ConstructorProperties 어노테이션 추가

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/interfaces/LoadingRepository.java:
```diff
@@ -0,0 +1,51 @@
+ * </p>
+ */
+public interface LoadingRepository {
+    /**
+     * 모든 Task 객체를 List의 형태로 불러온다
+     * <p>
+     * Method Returns:
+     * List<Task> - Task 엔티티 객체를 내부 요소로 하는 List Collection 객체
+     * </p>
+     */
```
```suggestion
    /**
     * 모든 Task를 리턴한다.
     * 
     * @return ,,,
     */
```

이런 형태가 되면 좋겠어요! 
javadoc에서 인자는 `@param <인자명> <설명>` 
리턴값은 `@return <설명>` 형태가 일반적이에요!

MJbae replied:
감사합니다. 참고해서 수정해보았습니다. 전보다 훨씬 깔끔해져서 기분이 좋습니다 !

MJbae committed:
chore: javadoc 컨벤션에 맞기 위해 수정

MJbae commented:
@ssisksl77 영환님 한 주 동안 리뷰 정말 감사합니다.
객체지향 패러다임에 대한 책을 읽고 관련 아이디어를 코드로 구현하자니 여러모로 어려움이 있었는데, 영환님의 코멘트를 보면서 혼자만의 생각에 갇히지 않고 다방면으로 고민할 수 있어서 좋았습니다!
Spring MVC에서 다루는 주요 객체의 역할에 맞추어 객체를 리팩토링하는 과정도 정말 좋습니다 :)

곧 또 코드로 뵐게요~
